lib/Default/MySqlDatabase.class.php

```diff
@@ -1,33 +1,43 @@
 <?php declare(strict_types=1);
 
-abstract class MySqlDatabase {
-	protected static $dbConn;
-	protected static $selectedDbName;
-	private static MySqlProperties $mysqlProperties;
-	protected $dbResult = null;
-	protected $dbRecord = null;
-
-	public function __construct() {
-		if (!self::$dbConn) {
-			// Set the mysqli driver to raise exceptions on errors
-			if (!mysqli_report(MYSQLI_REPORT_ERROR | MYSQLI_REPORT_STRICT)) {
-				$this->error('Failed to enable mysqli error reporting');
-			}
-			$mysqlProperties = new MySqlProperties();
-			self::$dbConn = new mysqli(
-				$mysqlProperties->getHost(),
-				$mysqlProperties->getUser(),
-				$mysqlProperties->getPassword(),
-				$mysqlProperties->getDatabaseName());
-			self::$mysqlProperties = $mysqlProperties;
-			self::$selectedDbName = $mysqlProperties->getDatabaseName();
-			// Default server charset should be set correctly. Using the default
-			// avoids the additional query involved in `set_charset`.
-			$charset = self::$dbConn->character_set_name();
-			if ($charset != 'utf8') {
-				$this->error('Unexpected charset: ' . $charset);
-			}
+use Smr\Container\DiContainer;
+use Smr\MySqlProperties;
+
+class MySqlDatabase {
+	private mysqli $dbConn;
+	private MySqlProperties $mysqlProperties;
+	private string $selectedDbName;
+	/**
+	 * @var bool | mysqli_result
+	 */
+	private $dbResult = null;
+	/**
+	 * @var array | null
+	 */
+	private $dbRecord = null;
+
+	public static function getInstance(bool $requireNewInstance = false): MySqlDatabase {
+		$class = MySqlDatabase::class;
+		return !$requireNewInstance
+			? DiContainer::get($class)
+			: DiContainer::make($class);
+	}
+
+	/**
+	 * MySqlDatabase constructor.
+	 * Not intended to be constructed by hand. If you need an instance of MySqlDatabase,
+	 * use MySqlDatabase::getInstance();
+	 * @param mysqli $dbConn The mysqli instance
+	 * @param MySqlProperties $mysqlProperties The properties object that was used to construct the mysqli instance
+	 */
+	public function __construct(mysqli $dbConn, MySqlProperties $mysqlProperties) {
+		$charset = $dbConn->character_set_name();
```

Comment:
The check on the character set is a sanity check on the mysqli object, not on the MySqlDatabase object, and it should only be performed once per connection. Therefore, it should probably be moved to the `mysqli::class` "maker" (what's the correct terminology for that?).

Reply:
The term is "factory".

I can move it to the factory and adjust the tests.

Reply:
Actually, now that I think about it, the charset check is outside the scope of the factory method. That check only takes place once the object is fully constructed. By that point, the factory has completed its responsibility. This means that this implementation detail, or business logic is the responsibility of the dependent who's using it -- the dependent determines whether or not they want to use the state of the dependency.

Factories should be dumb, and just "give you a thing". They shouldn't contain nearly any business logic. An exception to this rule is if you need some business logic to determine HOW to create a dependency, or if a dependency is even eligible for construction. For instance, this is valid factory business logic:

```php
if (!mysqli_report(MYSQLI_REPORT_ERROR | MYSQLI_REPORT_STRICT)) {
	throw new RuntimeException('Failed to enable mysqli error reporting');
}
```

Reply:
How would you test that branch of logic if it happened in the factory method?

Reply:
I would test it by mocking the `mysqli` constructor.

Reply:
How can you mock something you don't have control of though? It's being constructed inside the method -- you don't have access to mock that guy's constructor at that point.

Reply:
Example:

Given this factory method:
```php
function factory(MySqlProperties $mysqlProperties): mysqli {
    if (!mysqli_report(MYSQLI_REPORT_ERROR | MYSQLI_REPORT_STRICT)) {
        throw new RuntimeException('Failed to enable mysqli error reporting');
    }
    $mysqli = new mysqli(
        $mysqlProperties->getHost(),
        $mysqlProperties->getUser(),
        $mysqlProperties->getPassword(),
        $mysqlProperties->getDatabaseName());
    //# You need to mock this call
    $charset = $mysqli->character_set_name();
    //# In order to test this branch
    if ($charset != 'utf8') {
        throw new RuntimeException('Unexpected charset: ' . $charset);
    }
    return $mysqli;
}
```

From this test function:
```php
function test_non_utf_throws_exception() {
    //# Given mysql properties mock
    $mysqlProperties = MockWWithAllPropertiesSet();
    //# When producing a mysqli object through the factory method
    $mysqli = factory($mysqlProperties); //# How can you mock mysqli? This instantiates a real object.
}
```

Reply:
Yep it's a good point. I guess something like [this](https://akrabat.com/replacing-a-built-in-php-function-when-testing-a-component/) (leveraging namespace resolution rules to override the global namespace)? In python I'd use a patch decorator, but it doesn't look like PHP(Unit) has that type of functionality.

Happy to go with the new design, and maybe in some iteration down the road there will be a way to get the best of both worlds.

Reply:
Mockery does have some support for [global mocks](http://docs.mockery.io/en/latest/reference/creating_test_doubles.html#overloading), and that was actually the first mocking framework I tried to use for testing `AbstractSmrAccount` without spinning up a full integration test. Unfortunately it has some drawbacks when needing to do stuff like this multiple times in the same test suite. It also required stuff like running `PHPUnit` tests in separate processes so namespace/class collisions wouldn't happen.

With that you lose the ability to set breakpoints in classes :( It was just not a good solution, but a solution that does indeed work when you have these types of "hard dependencies" being constructed in the middle of your test subject methods, and have no alternative.